os/drivers/input/ist415.c

```diff
@@ -89,185 +98,1910 @@ struct touchscreen_ops_s g_ist415_ops = {
- *   as /dev/touchN where N is determined by the minor device number.
- *
- * Input Parameters:
- *   i2c_dev - The i2c device structure for the I2C bus that the touch controller is connected to
- *   config - The lower half touch driver structure that contains the touch configuration information.
- *
- * Returned Value:
- *   On success, a non-NULL pointer to the touch device structure touch_dev_s is returned.
- *   On a failure, NULL is returned.
- *
+ * Name: ist415_isp_erase
  ****************************************************************************/
 
-int ist415_initialize(const char*path, struct ist415_dev_s *priv)
+static int ist415_isp_erase(struct ist415_dev_s *dev, uint8_t target, uint8_t area, uint32_t addr, int size)
 {
 	int ret = 0;
-	uint8_t reg[1];
-	priv->ops->irq_disable();
+	int count = 0;
+	int unit = 0;
 
-	reg[0] = 0x23;
-	ret = i2c_write(priv->i2c, &(priv->i2c_config), (uint8_t *)reg, 1);
-	if (ret < OK) {
-		touchdbg("ERROR: i2c_write failed\n");
-		return NULL;
+	touchvdbg("%s\n", __func__);
+
+	if (area == ISP_CMD_MASS_ERASE) {
+		count = 1;
+	} else if (area == ISP_CMD_SECTOR_ERASE) {
+		count = size / IST415_ROM_SECTOR_SIZE;
+		unit = IST415_ROM_SECTOR_SIZE;
+	} else if (area == ISP_CMD_PAGE_ERASE) {
+		count = size / IST415_ROM_PAGE_SIZE;
+		unit = IST415_ROM_PAGE_SIZE;
+	}
+
+	for (int i = 0; i < count; i++) {
+		ret = ist415_i2c_write_isp(dev, target, area, addr + (i * unit), NULL, 0);
+		if (ret < 0) {
+			touchdbg("ERROR: ISP erase fail\n");
+		}
+	}
+	
+	return ret;
+}
+
+/****************************************************************************
+ * Name: ist415_isp_program
+ ****************************************************************************/
+
+static int ist415_isp_program(struct ist415_dev_s *dev, uint8_t target, uint32_t addr, uint32_t *buf32, int size)
+{
+	int ret = 0;
+	int len = size / sizeof(uint32_t);
+	
+	touchvdbg("isp program\n");
+
+	ret = ist415_i2c_write_isp(dev, target, ISP_CMD_PROG, addr, buf32, len);
+	if (ret < 0) {
+		touchdbg("ERROR: ISP write fail\n");
+	}
+
+	return ret;
+}
+
+
+/****************************************************************************
+ * Name: ist415_isp_fw_update
+ ****************************************************************************/
+
+static int ist415_isp_fw_update(struct ist415_dev_s *dev, const uint8_t *buf)
+{
+	int ret = 0;
+	FAR struct touchscreen_s *upper = dev->upper;
+
+	touchvdbg("isp firmware update\n");
+
+	touch_reset(upper, true);
+
+	ret = ist415_isp_enable(dev, true);
+	if (ret < 0) {
+		touch_reset(upper, false);
+		return ret;
+	}
+
+	ret = ist415_isp_erase(dev, ISP_TARGET_MAIN, ISP_CMD_SECTOR_ERASE, IST415_ROM_BASE_ADDR, IST415_ROM_MAIN_SIZE);
+	if (ret < 0) {
+		touch_reset(upper, false);
+		return ret;
+	}
+
+	// TODO: Delay 1ms
+	usleep(1000);
+
+	ret = ist415_isp_program(dev, ISP_TARGET_MAIN, IST415_ROM_BASE_ADDR, (uint32_t *)buf, IST415_ROM_MAIN_SIZE);
+	if (ret < 0) {
+		touch_reset(upper, false);
```

Comment:
All fail/return cases(duplicate) can be combined to call goto statement at function end